tests/NgTests/Validators/RegistrationValidatorTests.cs

```diff
@@ -0,0 +1,123 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Threading.Tasks;
+using NuGet.Services.Metadata.Catalog.Monitoring;
+using Xunit;
+
+namespace NgTests
+{
+    public class RegistrationValidatorTests
+    {
+        public static IEnumerable<object[]> ValidatorEqualIndexTestData => ValidatorTestData(t => GetEqualPairs(t.CreatePossibleIndexes));
+
+        public static IEnumerable<object[]> ValidatorUnequalIndexTestData => ValidatorTestData(t => GetUnequalPairs(t.CreatePossibleIndexes));
+
+        public static IEnumerable<object[]> ValidatorEqualLeafTestData => ValidatorTestData(t => GetEqualPairs(t.CreatePossibleLeafs));
+
+        public static IEnumerable<object[]> ValidatorUnequalLeafTestData => ValidatorTestData(t => GetUnequalPairs(t.CreatePossibleLeafs));
+
+        private static IEnumerable<object[]> ValidatorTestData<T>(Func<RegistrationValidatorTestData, IEnumerable<Tuple<T, T>>> getPairs)
+        {
+            foreach (var testData in RegistrationValidatorTestDatas)
+            {
+                var validator = testData.CreateValidator();
+
+                foreach (var pair in getPairs(testData))
+                {
+                    yield return new object[]
+                    {
+                            validator,
+                            pair.Item1,
+                            pair.Item2
+                    };
+                }
+            }
+        }
+
+        private static IEnumerable<RegistrationValidatorTestData> RegistrationValidatorTestDatas
+        {
+            get
+            {
+                var registrationValidatorTestDataTypes = 
+                    Assembly.GetExecutingAssembly().GetTypes()
+                        .Where(p =>
+                            typeof(RegistrationValidatorTestData)
+                            .IsAssignableFrom(p)
+                            && !p.IsAbstract);
+
+                return registrationValidatorTestDataTypes.Select(
+                    t => t.GetConstructor(new Type[] { }).Invoke(null) as RegistrationValidatorTestData);
+            }
+        }
+
+        private static IEnumerable<Tuple<T, T>> GetEqualPairs<T>(IEnumerable<T> values)
+        {
+            foreach (var obj in values)
+            {
+                yield return Tuple.Create(obj, obj);
```

Comment:
This should create 2 objects that have the same values, and not a tuple of two references to the same object. That way, we actually test your comparisons.